### ISSUE TITLE
Unexpected incompatibility of inline dict with container of `TypedDict`s, with `reportUnnecessaryContains` enabled

### ISSUE THREAD

**Issue by kylebebak** [microsoft/pyright#6436]
In certain contexts, e.g. with function calls, pyright checks whether an inline dict argument is assignable to a `TypedDict` parameter. But with `reportUnnecessaryContains` enabled it can't perform the same check for an inline dict and a container of `TypedDict`s.

```python
from typing import TypedDict


class FooDict(TypedDict):
    foo: str


def fn(d: FooDict):
    pass


# Ok, pyright sees that {"foo": "bar"} is assignable to FooDict in this function call
fn({"foo": "bar"})


b: list[FooDict] = []
# If reportUnnecessaryContains enabled, pyright unexpectedly raises error: `types "dict[str, str]" and "FooDict" have no overlap`
if {"foo": "bar"} in b:
    pass


c: FooDict = {"foo": "bar"}
# pyright raises no error, because we've told it that c has type of FooDict
if c in b:
    pass
```

Running pyright `1.1.335` from the command line

**Comment by erictraut** [comment]
Thanks for the reporting the issue. The `reportUnnecessaryContains` check is not foolproof. It employs various heuristics, and they can fail. This is the main reason `reportUnnecessaryContains` is not enabled by default in basic type checking mode. You may need to make adjustments to your code in some cases to avoid false positives with this check.

I'll think a bit more about whether this particular case can be improved.

**Comment by kylebebak** [comment]
Understood, and thank you for the explanation!

**Comment by erictraut** [comment]
I've added some extra special-case logic to handle this particular case. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.336, which I just published. It will also be included in a future release of pylance.
